Version 3

skpy opened the pull request:
This is a pretty big reworking to account for on-going Sentry development.

The biggest changes are:
- add new `sentry::config` and `sentry::setup` classes
- make all sub-classes private
- better manage class containment and ordering (Closes #22)
- restart background services after upgrade (Closes #21) 
- better bootstrap on very first install
- make default team the same name as the default organization
- initial spec test -- much more to come!

Due to changes in the upstream Sentry code and how it handles configuration, this module will not be able to install older versions of Sentry.  The default version to install is `latest`, which at this time is [8.4.1](https://pypi.python.org/pypi/sentry/8.4.1).  Arbitrary versions can be installed, but I have only tested it with [8.4.0].

Upgrades from older version of Sentry managed by older versions of this module should work just fine, provided both happen in tandem.  That is to say: using puppet-sentry 2.0.1 with Sentry 8.0.0, you can upgrade to puppet-sentry 3.0.0 and Sentry 8.4.1 simultaneously.  Other combinations have not been tested at this time.

Commit:
refactor for version 2 release

Commit:
Prep for v3 release

* Add new setup and config class
* Define resource ordering better
* Properly bootstrap a fresh install
* Remove `team` param: use org name for team name
* Update metadata: bump version, list dependencies
* Add stupid spec test for later refinement
* Manage `sentry beat` service
* Restart services on version change

dansajner commented on manifests/init.pp:
```diff
@@ -113,8 +110,46 @@
   $wsgi_threads      = $sentry::params::wsgi_threads,
 ) inherits ::sentry::params {
 
+  class { '::sentry::setup':
```
Can these class declarations be removed in favor of the `contain` statements plus chaining arrows?

skpy replied:
I think we can, now that I've made all the sub-classes private.  I'll try, and push a change if it's successful.

skpy replied:
This isn't proving quite as straightforward as I expected.  More work is needed.  I'd rather not delay this merge just for this tweak.

dansajner commented on Gemfile:
```diff
@@ -0,0 +1,7 @@
+source "https://gems.covermymeds.com"
```
This needs to by rubygems.org

dansajner commented on README.md:
```diff
@@ -106,55 +108,31 @@ Class parameters:
-* **ssl_* **: Apache SSL controls
-* **team**: name of the default team to create and use for new projects
-* **url**: source URL form which to install Sentry (false, use PyPI)
+* **ssl_* **: Apache SSL controls.  The `ssl_cert` and `ssl_key` parameters reference files on the Sentry server.  You are responsible for getting these files onto the server yourself.
```
Maybe change the wording here to "Your sentry profile class or other higher level class should manage the installation of these file resources.".

dansajner commented on spec/classes/init_spec.rb:
```diff
@@ -0,0 +1,38 @@
+require 'spec_helper'
+describe 'Sentry' do
+  let(:facts) do
```
Is there a reason for including this set of facts?

skpy replied:
This was to allow me to run the spec tests on my laptop, which is not running RHEL7.  Is there a better way to support running spec tests on platforms other than the target platform?

dansajner replied:
No,  I didn't realize the module has dependencies on os level facts.

skpy replied:
The puppetlabs-apache module splits on, at least, the `operatingsystemrelease` fact:

```
Puppet::PreformattedError:
       Evaluation Error: Error while evaluating a Function Call, 'split' parameter 'str' expects a String value, got Undef at .../modules/sentry/spec/fixtures/modules/apache/manifests/version.pp:7:16
```

And we require the custom `python_version` fact exposed from stankevich-python in our `sentry::wsgi` class:

```
     Puppet::PreformattedError:
       Evaluation Error: Error while evaluating a Function Call, 'regsubst' parameter 'target' expects a value of type Array or String, got Undef at .../modules/sentry/spec/fixtures/modules/sentry/manifests/wsgi.pp:59:13
```

The included facts are more comprehensive than what's actually needed, though. I'll pare it down.  It looks like all we need are:

```
osfamily: 'RedHat',
operatingsystem: 'RedHat',
operatingsystemrelease: '7.2',
python_version: '2.7.5'
```

dansajner commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,40 @@
+  c.hiera_config = 'spec/fixtures/hiera/hiera.yaml'
+end
+
+shared_context 'RedHat 7' do
```
I don't see these facts being used.  Is there a reason for including them?

skpy replied:
Nope. That was a mistake.

dansajner commented on manifests/service.pp:
```diff
@@ -49,4 +49,66 @@
+    notify  => Exec['enable-sentry-services'],
+  }
+
+  service { 'sentry-beat':
```
We're currently launching the beat process from the main worker service by passing the `-B` argument: https://github.com/covermymeds/puppet-sentry/blob/master/templates/sentry-worker.service.erb#L13.  Need to reconcile this as running more than 1 beat sounds bad.

```
/var/lib/sentry/bin/celery worker --help
Usage: celery worker [options] 
...
...
  -B, --beat            Also run the celery beat periodic task scheduler.
                        Please note that there must only be one instance of
                        this service.
...
...
```

skpy replied:
Great catch!

The [official Sentry documentation](https://docs.getsentry.com/on-premise/server/queue/) references running each process separately, and makes no mention of using the `-B` flag.  I honestly don't know the ramifications of either options.

Any preferences?  I can yank out the explicit beat stuff from this PR, or I can remove the `-B` flag from the worker invocation.

dansajner replied:
At one point the documentation mentioned the `-B` option which is why we've been using it :)  I suppose I prefer the current worker configuration as it keeps the puppet module a bit less complex.

Commit:
Address Dan's comments

* use rubygems.org in Gemfile
* update language in README
* remove unnecessary facts from spec_helper.rb
* rename spec/classes/init_spec.rb to spec/classes/sentry_spec.rb
* trim down list of facts defined in spec/classes/sentry_spec.rb
* add a spec test to confirm failure with Sentry version < 8.4.0
* update CHANGELOG

Commit:
typo in sentry::config param list

Commit:
Use resource containment, ftw

By using private classes, we can ensure that the parameters passed into
them will be those set for the top-level `sentry` class itself.  This,
in turn, means we don't need to use resource-like class invocaitons.  We
just need to ensure that resource ordering is properly defined, and that
earlier classes notify later classes.

skpy commented:
@dansajner @nchowning I've pushed another change, after successfully testing the include-like containment solution.  I was thrown off last week by a typo in one of the class parameters!  Once this was fixed, testing worked as I expected.

Please review!

dansajner commented on CHANGELOG:
```diff
@@ -1,3 +1,12 @@
+- better bootstrap on very first install
+- make default team the same name as the default organization
+- only support Sentry versions 8.4.0 and up, due to changes to Sentry code
+initial spec test
```
Missed a hyphen

Commit:
Remove -B from sentry-worker command line

Update CHANGELOG to add missing hyphen

skpy commented:
I think all outstanding issues have been resolved, and this is ready to merge.

I have elected to manage both the `sentry-worker` and `sentry-beat` services, since the documentation explicitly describes [background workers](https://docs.getsentry.com/on-premise/server/installation/#starting-background-workers) and [the beat cron process](https://docs.getsentry.com/on-premise/server/installation/#starting-the-cron-process).

I have performed an upgrade from an 8.1.2 install to 8.4.1, with a ~7GB database. The database migrations ran automatically just fine.

Once we merge, we can update our own installation from the git master branch.  When that proves successful, we can cut a new Puppet Forge release.  In the event that additional tweaks are necessary, we can make them in master before making the Forge release, so as to minimize iteration on that process.

dansajner commented:
👍  to merge.  🏆  for @skpy
